cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/fsm/DFA.kt

```diff
@@ -0,0 +1,156 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.analysis.fsm
+
+import de.fraunhofer.aisec.cpg.graph.Node
+
+/** A representation of a deterministic finite automaton (DFA). */
+class DFA : Cloneable {
+    companion object {
+        @JvmStatic val EPSILON: String = "ε"
+    }
+
+    var states = mutableSetOf<State>()
+    private var stateCounter = 1
+    var currentState: State? = null
+    var executionTrace = mutableListOf<Triple<State, Node, BaseOpEdge>>()
+
+    /**
+     * Creates an edge between two nodes with a given label (operator and optional base).
+     *
+     * It checks if [from] already has an outgoing edge with the same [base] and [op] but to another
+     * target node. If so, it does not add the edge end returns `false`, otherwise it creates the
+     * edge and returns `true`.
+     */
+    fun addEdge(from: State, to: State, op: String, base: String?): Boolean {
+        if (!states.contains(from)) {
+            states.add(from)
+        }
+        if (!states.contains(to)) {
+            states.add(to)
+        }
+        if (from.outgoingEdges.any { e -> e.matches(base, op) && e.nextState != to }) {
+            throw FSMBuilderException(
+                "State already has an outgoing edge with the same label but a different target!"
+            )
+        }
+        from.addOutgoingEdge(BaseOpEdge(op, base, to))
+        return true
+    }
+
+    /** Generates a new state and adds it to this FSM. */
+    fun addState(isStart: Boolean = false, isAcceptingState: Boolean = false): State {
+        val newState = State("q$stateCounter", isStart, isAcceptingState)
+        if (isStart) {
+            currentState = newState
+        }
+        states.add(newState)
+        stateCounter++
+        return newState
+    }
+
+    /**
+     * Generates the string representing this FSM in DOT format. This allows a simple visualization
+     * of the resulting automaton.
+     */
+    override fun toString(): String {
```

Comment:
is it feasible to have this as the standard `toString`? or should we have something like a `toDotString`?

Reply:
Changed the method name to `toDotString()`. In my opinion, it's easier to check if ones automaton is constructed correctly with such a graphical representation.